File: src/values/_type.rs
use crate::bytecodes::ApicaTypeBytecode;
use crate::values::bool::ValueBool;
use crate::values::string::ValueString;
use crate::values::value::Value;

#[derive(Clone)]
pub struct ValueType {
    kind: Option<ApicaTypeBytecode>,
    contained: Option<Vec<ValueType>>,
}

pub fn get_kind_repr(kind: &ApicaTypeBytecode) -> &'static str {
    match kind {
        ApicaTypeBytecode::Null => "null",
        ApicaTypeBytecode::Any => "any",
        ApicaTypeBytecode::I8 => "i8",
        ApicaTypeBytecode::I16 => "i16",
        ApicaTypeBytecode::I32 => "i32",
        ApicaTypeBytecode::I64 => "i64",
        ApicaTypeBytecode::U8 => "u8",
        ApicaTypeBytecode::U16 => "u16",
        ApicaTypeBytecode::U32 => "u32",
        ApicaTypeBytecode::U64 => "u64",
        ApicaTypeBytecode::F32 => "f32",
        ApicaTypeBytecode::F64 => "f64",
        ApicaTypeBytecode::Bool => "bool",
        ApicaTypeBytecode::Char => "char",
        ApicaTypeBytecode::String => "string",
        ApicaTypeBytecode::Error => "error",
        ApicaTypeBytecode::Type => "type",
    }
}

impl ValueType {
    pub fn init_empty() -> ValueType {
        ValueType { kind: None, contained: None }
    }

    pub fn init_with(kind: ApicaTypeBytecode, contained: Option<Vec<ValueType>>) -> ValueType {
        ValueType { kind: Some(kind), contained }
    }
    
    pub fn show(&self, end: char) {
        if let Some(kind) = &self.kind {
            let repr = get_kind_repr(kind);
            print!("{repr}");
            if let Some(contained) = &self.contained {
                print!("<");
                for i in 0..contained.len() {
                    let c = &contained[i];
                    c.show('\0');
                    if i != contained.len() - 1 {
                        print!(", ");
                    }
                }

                print!(">");
            }
        } else {
            print!("null{end}");
        }
    }
    
    pub fn is_null(&self) -> bool {
        self.kind.is_none()
    }
    
    pub fn get_type_representation(&self) -> &str {
        "type"
    }

    pub fn get_kind(&self) -> &Option<ApicaTypeBytecode> {
        &self.kind
    }

    pub fn get_contained(&self) -> &Option<Vec<ValueType>> {
        &self.contained
    }

    pub fn to_string(&self) -> String {
        if let Some(kind) = &self.kind {
            let mut string = format!("{}", get_kind_repr(kind));
            if let Some(contained) = &self.contained {
                string.push('<');
                for c in contained {
                    let additional = c.to_string();
                    if !additional.is_empty() {
                        string.push_str(additional.as_str());
                    }
                }
                string.push('>');
            }

            string
        } else {
            String::new()
        }
    }

    pub fn not(&self) -> Value {
        Value::Bool(ValueBool::init_with(!self.is_null()))
    }
    
    pub fn convert(&self, to: ApicaTypeBytecode) -> Option<Value> {
        if let Some(_) = &self.kind {
            match to {
                ApicaTypeBytecode::Bool => Some(Value::Bool(ValueBool::init_with(true))),

                ApicaTypeBytecode::String => Some(Value::String(ValueString::init_with(self.to_string()))),

                _ => None,
            }
        } else {
            match to {
                ApicaTypeBytecode::Bool => Some(Value::Bool(ValueBool::init_empty())),

                ApicaTypeBytecode::String => Some(Value::String(ValueString::init_empty())),

                _ => None,
            }
        }
    }

    pub fn auto_convert(&self, to: ApicaTypeBytecode) -> Option<Value> {
        if let Some(_) = &self.kind {
            match to {
                ApicaTypeBytecode::Any => Some(Value::Type(self.clone())),
                
                ApicaTypeBytecode::Type => Some(Value::Type(ValueType::init_with(ApicaTypeBytecode::Type, None))),

                _ => None,
            }
        } else {
            match to {
                ApicaTypeBytecode::Any => Some(Value::Type(ValueType::init_empty())),
                
                ApicaTypeBytecode::Type => Some(Value::Type(ValueType::init_with(ApicaTypeBytecode::Type, None))),

                _ => None,
            }
        }
    }
}

File: src/values/type.cpp
#include "values/type.hpp"
#include "values/bool.hpp"
#include "values/string.hpp"
#include "elements.hpp"

using namespace common::values;

ValueType::ValueType()
    : type_kind(std::nullopt) {
    
}

ValueType::ValueType(common::bytecodes::ApicaTypeBytecode type_kind)
    : type_kind(type_kind) {
    
}

std::string ValueType::getKindRepr(common::bytecodes::ApicaTypeBytecode kind) {
    switch (kind) {
        case common::bytecodes::ApicaTypeBytecode::Null: return "null";

        case common::bytecodes::ApicaTypeBytecode::I8: return "i8";
        case common::bytecodes::ApicaTypeBytecode::I16: return "i16";
        case common::bytecodes::ApicaTypeBytecode::I32: return "i32";
        case common::bytecodes::ApicaTypeBytecode::I64: return "i64";
        case common::bytecodes::ApicaTypeBytecode::U8: return "u8";
        case common::bytecodes::ApicaTypeBytecode::U16: return "u16";
        case common::bytecodes::ApicaTypeBytecode::U32: return "u32";
        case common::bytecodes::ApicaTypeBytecode::U64: return "u64";
        case common::bytecodes::ApicaTypeBytecode::F32: return "f32";
        case common::bytecodes::ApicaTypeBytecode::F64: return "f64";
        case common::bytecodes::ApicaTypeBytecode::Bool: return "bool";

        case common::bytecodes::ApicaTypeBytecode::Char: return "char";
        case common::bytecodes::ApicaTypeBytecode::String: return "string";

        case common::bytecodes::ApicaTypeBytecode::Error: return "error";
        case common::bytecodes::ApicaTypeBytecode::Type: return "type";

        default: return "???";
    }
}

void ValueType::show(char end) const {
    if (this->type_kind) {
        std::cout << "type<" << ValueType::getKindRepr(this->type_kind.value()) << '>' << end;
    } else {
        std::cout << "type<null>" << end;
    }
}

bool ValueType::isNull() const {
    return !this->type_kind.has_value();
}

std::string ValueType::getTypeRepr() const {
    return "type";
}

common::bytecodes::ApicaTypeBytecode ValueType::getKind() const {
    return common::bytecodes::ApicaTypeBytecode::Type;
}

std::optional<Value*> ValueType::add(const Value *) const {
    return std::nullopt;
}

std::optional<Value*> ValueType::increment() {
    return std::nullopt;
}

std::optional<Value*> ValueType::leftIncrement() {
    return std::nullopt;
}

std::optional<Value*> ValueType::subtract(const Value *) const {
    return std::nullopt;
}

std::optional<Value*> ValueType::decrement() {
    return std::nullopt;
}

std::optional<Value*> ValueType::leftDecrement() {
    return std::nullopt;
}

std::optional<Value*> ValueType::unaryNot() const {
    return new ValueBool(this->type_kind.has_value() ? this->type_kind.value() == common::bytecodes::ApicaTypeBytecode::Null : true);
}

std::optional<Value*> ValueType::bitwiseNot() const {
    return std::nullopt;
}

std::optional<Value*> ValueType::convert(common::bytecodes::ApicaTypeBytecode to) const {
    if (this->type_kind) {
        switch (to) {
            case common::bytecodes::ApicaTypeBytecode::Bool:
                return new ValueBool(this->type_kind.value() != common::bytecodes::ApicaTypeBytecode::Null);
            
            case common::bytecodes::ApicaTypeBytecode::String: {
                std::string result("<");
                result += ValueType::getKindRepr(this->type_kind.value());
                result += '>';

                return new ValueString(result);
            }

            default: return std::nullopt;
        }
    } else {
        switch (to) {
            case common::bytecodes::ApicaTypeBytecode::Bool:
                return new ValueBool();
            
            case common::bytecodes::ApicaTypeBytecode::String:
                return new ValueString();
            
            default: return std::nullopt;
        }
    }
}

std::optional<Value*> ValueType::autoConvert(common::bytecodes::ApicaTypeBytecode to) const {
    if (this->type_kind) {
        switch (to) {
            case common::bytecodes::ApicaTypeBytecode::Any:
                return new ValueType(this->type_kind.value());
            
            case common::bytecodes::ApicaTypeBytecode::Type:
                return new ValueType(common::bytecodes::ApicaTypeBytecode::Type);

            default: return std::nullopt;
        }
    } else {
        switch (to) {
            case common::bytecodes::ApicaTypeBytecode::Any:
                return new ValueType();
            
            case common::bytecodes::ApicaTypeBytecode::Type:
                return new ValueType(common::bytecodes::ApicaTypeBytecode::Type);

            default: return std::nullopt;
        }
    }
}

std::optional<common::bytecodes::ApicaTypeBytecode> ValueType::getTypeKind() const {
    return this->type_kind;
}